## Задача 1. My heart will go on

Датасет **titanic** из библиотеки `Seaborn` содержит информацию о пассажирах легендарного корабля Титаник, который затонул в 1912 году после столкновения с айсбергом. Этот набор данных часто используется для обучения и тестирования алгоритмов машинного обучения, особенно в задачах бинарной классификации (выжил / не выжил).

**Описание данных**

| Поле         | Тип      | Описание |
|--------------|----------|----------|
| `survived`   | int      | Выжил (1) или не выжил (0) |
| `pclass`     | int      | Класс билета (1, 2, 3) |
| `sex`        | str      | Пол (`male`/`female`) |
| `age`        | float    | Возраст |
| `sibsp`      | int      | Количество братьев/сестёр/супругов на борту |
| `parch`      | int      | Количество родителей/детей на борту |
| `fare`       | float    | Стоимость билета |
| `embarked`   | str      | Порт посадки (`C`=Cherbourg, `Q`=Queenstown, `S`=Southampton) |
| `class`      | str      | Класс билета (`First`, `Second`, `Third`) |
| `who`        | str      | Категория: `man`, `woman` или `child` |
| `adult_male` | bool     | Является ли взрослым мужчиной |
| `deck`       | str      | Палуба |
| `embark_town`| str      | Название порта посадки |
| `alive`      | str      | Выжил (`yes`/`no`) |
| `alone`      | bool     | Путешествовал один |

**Загрузка датасета**

In [154]:
import seaborn as sns
import numpy as np
import pandas as pd


titanic_data = sns.load_dataset("titanic")
titanic_data.sample(5)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
259,1,2,female,50.0,0,1,26.0000,S,Second,woman,False,NaN,Southampton,yes,False
395,0,3,male,22.0,0,0,7.7958,S,Third,man,True,NaN,Southampton,no,True
278,0,3,male,7.0,4,1,29.1250,Q,Third,child,False,NaN,Queenstown,no,False
782,0,1,male,29.0,0,0,30.0000,S,First,man,True,D,Southampton,no,True
512,1,1,male,36.0,0,0,26.2875,S,First,man,True,E,Southampton,yes,True


**Задача**

Ниже описаны 10 небольших заданий, которые вам необходимо решить.

**Подсказка**:

В некоторых заданиях вам может быть полезен метод `value_counts`.

### Часть 1

Определите число пропущенных данных для каждого столбца таблицы `titanic_data`.

In [155]:
titanic_data.isnull().sum()


survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

### Часть 2

Удалите все столбцы, количество пропусков в которых превышает половину количества строк в таблице.

После того, как вы удалите все столбцы, нарушающие описанное условие, удалите все строки, количество пропусков в которых превышает половину количества столбцов.

In [158]:
titanic_data = titanic_data.loc[:, titanic_data.isnull().mean() <= 0.5]
# Удаляем строки, где NaN > 50% от количества столбцов
titanic_data = titanic_data.loc[titanic_data.isnull().mean(axis=1) <= 0.5]
titanic_data.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
embark_town      2
alive            0
alone            0
dtype: int64

### Часть 3

Если вы сделали все правильно, больше всего пропусков должно остаться в столбце `"age"` - 177. Их необходимо заполнить. Заполним пропуски следующим образом:
- Если значение столбца `"who"="man"`, пропуск необходимо заполнить медианным значением известных возрастов мужчин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="woman"`, пропуск необходимо заполнить медианным значением известных возрастов женщин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="child"`, пропуск необходимо заполнить медианным значением известных возрастов детей, округленным до ближайшего целого числа;

In [159]:
median_man_age = titanic_data[titanic_data["who"] == "man"]["age"].median()
mask_man = titanic_data["who"] == "man"
titanic_data.loc[mask_man, "age"] = titanic_data.loc[mask_man, "age"].fillna(median_man_age)

median_woman_age = titanic_data[titanic_data["who"] == "woman"]["age"].median()
mask_woman = titanic_data["who"] == "woman"
titanic_data.loc[mask_woman, "age"] = titanic_data.loc[mask_woman, "age"].fillna(median_woman_age)

median_child_age = titanic_data[titanic_data["who"] == "child"]["age"].median()
mask_child = titanic_data["who"] == "child"
titanic_data.loc[mask_child, "age"] = titanic_data.loc[mask_child, "age"].fillna(median_child_age)

### Часть 4

Удалите все строки, в которых осталось больше одного пропуска. Если вы все сделали правильно, после этого действия в таблице не должно остаться пропусков.

In [160]:
titanic_data = titanic_data[titanic_data.isnull().sum(axis=1) <= 1]   

### Часть 5

Определите название города, из которого отправилось больше всего пассажиров.

In [161]:
titanic_data["embark_town"].value_counts().index[0]


'Southampton'

### Часть 6

Определите процент выживших пассажиров от числа пассажиров, оставшихся в таблице после очистки данных. Ответ округлите до 2 знаков после запятой.

In [162]:
round(titanic_data["alive"].value_counts(normalize=True)["yes"] * 100, 3)

np.float64(38.245)

### Часть 7

Определите число выживших пассажиров для каждого пункта отправления. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия пунктов отправления, а значения - число выживших пассажиров.

In [163]:
titanic_data[titanic_data['alive'] == 'yes']['embark_town'].value_counts()

embark_town
Southampton    217
Cherbourg       93
Queenstown      30
Name: count, dtype: int64

### Часть 8

Определите процент выживших пассажиров в каждом классе. Значения округлите до 2 знаков после запятой. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия классов, а значения - процент выживших пассажиров.

In [164]:
titanic_data[titanic_data['alive'] == 'yes']['class'].value_counts(normalize=True)

class
First     0.394118
Third     0.350000
Second    0.255882
Name: proportion, dtype: float64

### Часть 9

Будем считать, что пассажиры, купившие билет **НЕ МЕНЕЕ** чем за $100, считаются богатыми. Определите процент выживших среди богатых пассажиров. Ответ округлите до 2 знаков после запятой. В ответе должно получиться число. 

In [165]:
round(titanic_data[titanic_data["fare"] >= 100]["alive"].value_counts(normalize=True)["yes"], 2)

np.float64(0.74)

### Часть 10

Определите количество детей, путешествовавших в одиночку.

In [167]:
titanic_data[(titanic_data["who"] == "child") & (titanic_data["alone"] == True)].shape[0]

6

Какие выводы вы можете сделать о выживших пассажирах Титаника? 

In [169]:
titanic_data[titanic_data["alive"] == "yes"]

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
8,1,3,female,27.0,0,2,11.1333,S,Third,woman,False,Southampton,yes,False
9,1,2,female,14.0,1,0,30.0708,C,Second,child,False,Cherbourg,yes,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
875,1,3,female,15.0,0,0,7.2250,C,Third,child,False,Cherbourg,yes,True
879,1,1,female,56.0,0,1,83.1583,C,First,woman,False,Cherbourg,yes,False
880,1,2,female,25.0,0,1,26.0000,S,Second,woman,False,Southampton,yes,False
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,Southampton,yes,True


![крутой миньон](image.png)